File: frontend/src/types/workspace.ts
export type GoalStatus = 'idle' | 'planning' | 'running' | 'paused' | 'waiting' | 'waiting_approval' | 'replanning' | 'revision_required' | 'blocked' | 'handoff' | 'reviewing' | 'completed' | 'failed' | 'cancelled';
export type TaskStatus = 'pending' | 'ready' | 'assigned' | 'running' | 'waiting_tool' | 'waiting_approval' | 'verifying' | 'replanning' | 'revision_required' | 'blocked' | 'completed' | 'completed_verified' | 'completed_unverified' | 'failed' | 'handoff' | 'skipped' | 'cancelled';
export type WorkerStatus = 'idle' | 'running' | 'handoff_required' | 'completed' | 'failed' | 'cancelled';

export interface Goal {
  id: string;
  title: string;
  description?: string | null;
  team_preset?: string | null;
  status: GoalStatus;
  execution_mode?: 'live' | 'sandbox' | 'dry_run' | 'mock';
  workspace_root?: string | null;
  final_verification_status?: string | null;
  budget_tokens?: number;
  budget_cost_usd?: number | null;
  max_duration_seconds?: number;
  max_parallel_tasks?: number;
  created_at?: string | null;
  updated_at?: string | null;
}

export interface WorkspaceTask {
  id: string;
  goal_id: string;
  title: string;
  description?: string | null;
  status: TaskStatus;
  assigned_agent_id?: string | null;
  assigned_worker_id?: string | null;
  output?: string | null;
  tokens_used: number;
  duration_ms?: number | null;
  priority: number;
  flow_position?: number;
  created_at?: string | null;
  updated_at?: string | null;
  agent_name?: string | null;
  agent_role?: string | null;
  model_name?: string | null;
  worker_status?: string | null;
  workspace_scope?: string | null;
  handoff?: WorkspaceHandoff | null;
  handoffs?: WorkspaceHandoff[];
  model_id?: string | null;
  quota?: WorkspaceQuota | null;
  routing_decision?: WorkspaceRoutingDecision | null;
  context?: string | null;
  recent_logs?: WorkspaceLog[];
  dependencies?: string[];
  acceptance_criteria?: Array<Record<string, unknown>>;
  verification_status?: string | null;
  artifacts?: Array<{ id: string; type: string; path?: string | null; checksum?: string | null; verification_status?: string | null }>;
  verification_results?: Array<{ id: string; criterion_type: string; command_or_rule: string; status: string; evidence?: string | null; exit_code?: number | null }>;
  context_runs?: ContextRun[];
  task_type?: string;
  blocked_reason?: string | null;
  changed_files?: string[];
  test_status?: string | null;
  latest_tool_call?: { id: string; tool_name: string; status: string; latency_ms: number; error_message?: string | null; result?: { changed_files?: string[] } } | null;
  recent_tool_calls?: Array<{ id: string; tool_name: string; status: string; latency_ms: number; error_message?: string | null; result?: { changed_files?: string[] } }>;
  plan_version_id?: string | null;
  plan_task_id?: string | null;
  plan_source?: string | null;
  selection_decision?: AgentSelectionDecision | null;
}

export interface PlanTask {
  id: string;
  client_task_id: string;
  objective: string;
  task_type: string;
  required_capabilities: string[];
  required_tools: string[];
  dependencies: string[];
  acceptance_criteria: Array<Record<string, unknown>>;
  risk_level: string;
  parallel_safe: boolean;
  context_query: string;
  approval_required: boolean;
  workspace_scope?: string | null;
  merge_strategy?: string | null;
  runtime_task_id?: string | null;
  source: string;
}

export interface ExecutionPlan {
  id: string;
  plan_id: string;
  goal_id: string;
  version: number;
  status: string;
  task_mode: 'direct' | 'single_agent' | 'sequential_multi_agent' | 'parallel_multi_agent';
  goal_summary: string;
  assumptions: string[];
  required_context: string[];
  activation_reason: string;
  final_acceptance_criteria: Array<Record<string, unknown>>;
  human_approval_points: string[];
  estimated_cost: Record<string, unknown>;
  fallback_reason?: string | null;
  planner_type: string;
  confirmed_at?: string | null;
  tasks: PlanTask[];
}

export interface PlanVersionSummary {
  id: string;
  plan_id: string;
  version: number;
  status: string;
  task_mode: string;
  activation_reason: string;
  planner_type: string;
  created_at?: string | null;
}

export interface PlanChange {
  id: string;
  goal_id: string;
  from_plan_version_id?: string | null;
  to_plan_version_id: string;
  change_type: string;
  reason: string;
  evidence: Array<Record<string, unknown>>;
  retained_task_ids: string[];
  cancelled_task_ids: string[];
  added_task_ids: string[];
  replaced_task_ids: string[];
  created_at?: string | null;
}

export interface WorkspaceQuota {
  model_id: string;
  quota_status: string;
  usage_percent?: number | null;
  estimated_remaining?: number | null;
  total_tokens: number;
  token_limit?: number | null;
  request_count: number;
}

export interface WorkspaceRoutingDecision {
  selected_model_id?: string;
  backup_model_ids?: string[];
  confidence?: number;
  routing_reason?: {
    summary?: string;
    primary_factors?: string[];
    secondary_factors?: string[];
    tradeoffs?: string[];
  };
  risk_flags?: Array<{ type: string; severity: string; message: string; suggestion?: string }>;
  score_breakdown?: Array<{
    model_id: string;
    model_name: string;
    total_score: number;
    dimension_scores: Array<{ dimension: string; score: number; weight: number; weighted_score: number; reason: string }>;
  }>;
  is_user_override?: boolean;
}

export interface WorkspaceLog {
  id: string;
  event_type: string;
  event_status: string;
  output_summary?: string | null;
  input_summary?: string | null;
  error_message?: string | null;
  quota_status?: string | null;
  workspace_scope?: string | null;
  model_id?: string | null;
  model_name?: string | null;
  agent_id?: string | null;
  agent_name?: string | null;
  created_at?: string | null;
}

export interface WorkspaceHandoff {
  id: string;
  task_id: string;
  status: string;
  reason: string;
  reason_description?: string | null;
  from_agent_id: string;
  from_agent_name?: string | null;
  from_model_id: string;
  to_agent_id: string;
  to_agent_name?: string | null;
  to_model_id: string;
  result_after_handoff?: string | null;
  created_at?: string | null;
  accepted_at?: string | null;
  completed_at?: string | null;
}

export interface WorkspaceAgent {
  id: string;
  name: string;
  role: string;
  status: string;
  default_model_id?: string | null;
  is_enabled: boolean;
}

export interface WorkspaceWorker {
  id: string;
  agent_id: string;
  model_id: string;
  goal_id?: string | null;
  task_id?: string | null;
  inherited_from_handoff_id?: string | null;
  status: string;
  total_tokens_used: number;
  model_name?: string | null;
  quota_status?: string | null;
  workspace_scope?: string | null;
  step_count?: number;
  failure_count?: number;
  last_observation?: string | null;
  next_action?: string | null;
}

export interface WorkspaceState {
  goal: Goal | null;
  tasks: WorkspaceTask[];
  agents: WorkspaceAgent[];
  workers: WorkspaceWorker[];
  handoffs: WorkspaceHandoff[];
  active_plan?: ExecutionPlan | null;
  plan_versions?: PlanVersionSummary[];
  task_mode?: ExecutionPlan['task_mode'] | null;
  activation_reason?: string | null;
  task_edges?: Array<{ source: string; target: string; source_client_task_id: string; target_client_task_id: string }>;
  parallel_groups?: Array<{ id: string; task_ids: string[] }>;
  replan_events?: PlanChange[];
  completion_evidence?: {
    allowed: boolean;
    status: string;
    goal_status: string;
    reason: string;
    failed_task_ids: string[];
    tasks: Array<{
      task_id: string;
      title: string;
      status: string;
      verification_status?: string | null;
      verification_required: boolean;
      satisfied: boolean;
      reasons: string[];
    }>;
  } | null;
  context_runs?: ContextRun[];
  selection_decisions?: AgentSelectionDecision[];
  multi_agent_metrics?: MultiAgentMetrics;
}

export interface AgentSelectionDecision {
  id: string;
  task_id: string;
  required_capabilities: string[];
  required_tools: string[];
  candidates: Array<{
    agent_id: string; agent_name: string; role: string; eligible: boolean;
    elimination_reasons: string[]; score: number; selected_model_id: string;
    selected_model_name: string; score_breakdown: Record<string, number>;
  }>;
  selected_agent_id: string;
  selected_model_id: string;
  backup_model_ids: string[];
  score: number;
  selection_reason: string;
  fallback_entry: { agent_ids?: string[]; model_ids?: string[]; handoff_allowed?: boolean; reason?: string };
  created_at?: string | null;
}

export interface MultiAgentMetrics {
  mode: ExecutionPlan['task_mode'];
  why_multi_agent: string;
  activated_agent_count: number;
  active_agent_count: number;
  active_parallelism: number;
  coordination_task_count: number;
  coordination_tokens: number;
  productive_tokens: number;
  coordination_token_ratio: number;
  coordination_duration_ms: number;
  parallel_task_count: number;
  single_agent_serial_baseline_ms: number;
  parallel_observed_estimate_ms: number;
  potential_parallel_saving_ms: number;
  estimated_net_time_benefit_ms: number;
  benefit_positive: boolean;
  measurement_note: string;
  mode_comparison: Record<'single_agent' | 'sequential_multi_agent' | 'parallel_multi_agent', { duration_ms: number; tokens: number; basis: string }>;
}

export interface ContextRun {
  id: string;
  goal_id?: string | null;
  task_id?: string | null;
  agent_id?: string | null;
  query: string;
  policy: string;
  filters: Record<string, unknown>;
  latency_ms: number;
  token_budget: number;
  token_count: number;
  status: string;
  created_at?: string | null;
  items: Array<{
    id: string;
    rank: number;
    score: number;
    used: boolean;
    citation?: string | null;
    token_count: number;
    source_id: string;
    source_name?: string | null;
    chunk_id?: string | null;
    path?: string | null;
    content?: string | null;
  }>;
}

export const GOAL_STATUS_LABELS: Record<string, string> = {
  idle: '未启动',
  planning: '规划中',
  running: '运行中',
  paused: '已暂停',
  waiting: '等待中',
  handoff: '交接中',
  reviewing: '审查中',
  completed: '已完成',
  completed_verified: '已验证完成',
  completed_unverified: '未验证完成',
  verifying: '验证中',
  revision_required: '需要修订',
  blocked: '已阻塞',
  failed: '失败',
  cancelled: '已停止',
};

export const GOAL_STATUS_COLORS: Record<string, string> = {
  idle: 'bg-stone-100 text-stone-600 border-stone-200',
  planning: 'bg-purple-100 text-purple-700 border-purple-200',
  running: 'bg-blue-100 text-blue-700 border-blue-200',
  paused: 'bg-stone-100 text-stone-700 border-stone-300',
  waiting: 'bg-amber-100 text-amber-700 border-amber-200',
  handoff: 'bg-purple-100 text-purple-700 border-purple-200',
  reviewing: 'bg-orange-100 text-orange-700 border-orange-200',
  completed: 'bg-green-100 text-green-700 border-green-200',
  failed: 'bg-red-100 text-red-700 border-red-200',
  cancelled: 'bg-stone-100 text-stone-600 border-stone-300',
};

export const TASK_STATUS_LABELS: Record<string, string> = {
  pending: '待处理',
  ready: '就绪',
  assigned: '已分配',
  running: '执行中',
  waiting_tool: '等待工具',
  waiting_approval: '等待审批',
  verifying: '验证中',
  revision_required: '需要修订',
  blocked: '已阻塞',
  completed: '已完成',
  completed_verified: '已验证完成',
  completed_unverified: '未验证完成',
  failed: '失败',
  handoff: '交接中',
  cancelled: '已取消',
};

export const TASK_STATUS_ICONS: Record<string, string> = {
  pending: '⏸',
  ready: '◌',
  assigned: '📋',
  running: '▶',
  waiting_tool: '⏳',
  waiting_approval: '⚠',
  verifying: '🔎',
  revision_required: '↩',
  blocked: '⛔',
  completed: '✓',
  completed_verified: '✓',
  completed_unverified: '⚠',
  failed: '✗',
  handoff: '🔄',
  cancelled: '—',
};

export const TASK_STATUS_BORDERS: Record<string, string> = {
  pending: 'border-dashed border-stone-300',
  assigned: 'border-stone-400',
  running: 'border-blue-500 shadow-blue-100 shadow-md',
  completed: 'border-green-500',
  completed_verified: 'border-green-500',
  completed_unverified: 'border-amber-500',
  verifying: 'border-blue-500',
  revision_required: 'border-amber-500',
  blocked: 'border-red-500',
  failed: 'border-red-500',
  handoff: 'border-purple-500 shadow-purple-100 shadow-md',
};

export const TASK_STATUS_BG: Record<string, string> = {
  pending: 'bg-white',
  assigned: 'bg-white',
  running: 'bg-blue-50',
  completed: 'bg-green-50',
  completed_verified: 'bg-green-50',
  completed_unverified: 'bg-amber-50',
  verifying: 'bg-blue-50',
  revision_required: 'bg-amber-50',
  blocked: 'bg-red-50',
  failed: 'bg-red-50',
  handoff: 'bg-purple-50',
};

export const WORKER_STATUS_COLORS: Record<string, string> = {
  idle: 'bg-stone-400',
  running: 'bg-blue-500',
  handoff_required: 'bg-purple-500',
  completed: 'bg-green-500',
  failed: 'bg-red-500',
};

File: backend/src/models/workspace.py
import json
import uuid
from datetime import datetime, timezone

from sqlalchemy import Boolean, Column, DateTime, Float, Integer, String, Text, UniqueConstraint, event

from src.core.database import Base


class Goal(Base):
    __tablename__ = "goals"

    id = Column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    title = Column(String(255), nullable=False)
    description = Column(Text, nullable=True)
    team_preset = Column(String(100), nullable=True, index=True)
    status = Column(String(50), nullable=False, default="idle", index=True)
    execution_mode = Column(String(20), nullable=False, default="live")
    workspace_root = Column(Text, nullable=True)
    run_id = Column(String(36), nullable=True, index=True)
    final_verification_status = Column(String(50), nullable=True)
    budget_tokens = Column(Integer, nullable=False, default=100000)
    budget_cost_usd = Column(Float, nullable=True)
    max_duration_seconds = Column(Integer, nullable=False, default=3600)
    max_parallel_tasks = Column(Integer, nullable=False, default=3)
    created_at = Column(DateTime, nullable=False, default=lambda: datetime.now(timezone.utc))
    updated_at = Column(DateTime, nullable=False, default=lambda: datetime.now(timezone.utc))

    def to_dict(self) -> dict:
        return {
            "id": self.id,
            "title": self.title,
            "description": self.description,
            "team_preset": self.team_preset,
            "status": self.status,
            "execution_mode": self.execution_mode,
            "workspace_root": self.workspace_root,
            "run_id": self.run_id,
            "final_verification_status": self.final_verification_status,
            "budget_tokens": self.budget_tokens,
            "budget_cost_usd": self.budget_cost_usd,
            "max_duration_seconds": self.max_duration_seconds,
            "max_parallel_tasks": self.max_parallel_tasks,
            "created_at": self.created_at.isoformat() if self.created_at else None,
            "updated_at": self.updated_at.isoformat() if self.updated_at else None,
        }


class Task(Base):
    __tablename__ = "tasks"

    id = Column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    goal_id = Column(String(36), nullable=False, index=True)
    title = Column(String(255), nullable=False)
    description = Column(Text, nullable=True)
    status = Column(String(50), nullable=False, default="pending", index=True)
    assigned_agent_id = Column(String(36), nullable=True, index=True)
    assigned_worker_id = Column(String(36), nullable=True)
    output = Column(Text, nullable=True)
    tokens_used = Column(Integer, nullable=False, default=0)
    duration_ms = Column(Integer, nullable=True)
    priority = Column(Integer, nullable=False, default=0)
    task_type = Column(String(50), nullable=False, default="general")
    required_capabilities = Column(Text, nullable=False, default="[]")
    required_tools = Column(Text, nullable=False, default="[]")
    dependencies = Column(Text, nullable=False, default="[]")
    acceptance_criteria = Column(Text, nullable=False, default="[]")
    risk_level = Column(String(20), nullable=False, default="low")
    retry_count = Column(Integer, nullable=False, default=0)
    max_retries = Column(Integer, nullable=False, default=2)
    verification_status = Column(String(50), nullable=True)
    blocked_reason = Column(Text, nullable=True)
    parent_task_id = Column(String(36), nullable=True, index=True)
    plan_version_id = Column(String(36), nullable=True, index=True)
    plan_task_id = Column(String(36), nullable=True, index=True)
    plan_source = Column(String(30), nullable=True)
    lease_expires_at = Column(DateTime, nullable=True, index=True)
    recovery_count = Column(Integer, nullable=False, default=0)
    created_at = Column(DateTime, nullable=False, default=lambda: datetime.now(timezone.utc))
    updated_at = Column(DateTime, nullable=False, default=lambda: datetime.now(timezone.utc))

    def to_dict(self) -> dict:
        return {
            "id": self.id,
            "goal_id": self.goal_id,
            "title": self.title,
            "description": self.description,
            "status": self.status,
            "assigned_agent_id": self.assigned_agent_id,
            "assigned_worker_id": self.assigned_worker_id,
            "output": self.output,
            "tokens_used": self.tokens_used,
            "duration_ms": self.duration_ms,
            "priority": self.priority,
            "task_type": self.task_type,
            "required_capabilities": self._get_json("required_capabilities"),
            "required_tools": self._get_json("required_tools"),
            "dependencies": self._get_json("dependencies"),
            "acceptance_criteria": self._get_json("acceptance_criteria"),
            "risk_level": self.risk_level,
            "retry_count": self.retry_count,
            "max_retries": self.max_retries,
            "verification_status": self.verification_status,
            "blocked_reason": self.blocked_reason,
            "parent_task_id": self.parent_task_id,
            "plan_version_id": self.plan_version_id,
            "plan_task_id": self.plan_task_id,
            "plan_source": self.plan_source,
            "lease_expires_at": self.lease_expires_at.isoformat() if self.lease_expires_at else None,
            "recovery_count": self.recovery_count,
            "created_at": self.created_at.isoformat() if self.created_at else None,
            "updated_at": self.updated_at.isoformat() if self.updated_at else None,
        }

    def _get_json(self, field: str):
        import json
        value = getattr(self, field, "[]")
        try:
            return json.loads(value or "[]")
        except json.JSONDecodeError:
            return []

    def set_json(self, field: str, value) -> None:
        import json
        setattr(self, field, json.dumps(value))


class ExecutionPlan(Base):
    """One immutable, validated version of a Goal's execution plan."""

    __tablename__ = "execution_plans"
    __table_args__ = (
        UniqueConstraint("goal_id", "version", name="uq_execution_plan_goal_version"),
        UniqueConstraint("plan_id", "version", name="uq_execution_plan_id_version"),
    )

    id = Column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    plan_id = Column(String(36), nullable=False, index=True)
    goal_id = Column(String(36), nullable=False, index=True)
    version = Column(Integer, nullable=False)
    status = Column(String(30), nullable=False, default="validated", index=True)
    task_mode = Column(String(40), nullable=False)
    goal_summary = Column(Text, nullable=False)
    assumptions = Column(Text, nullable=False, default="[]")
    required_context = Column(Text, nullable=False, default="[]")
    activation_reason = Column(Text, nullable=False)
    final_acceptance_criteria = Column(Text, nullable=False, default="[]")
    human_approval_points = Column(Text, nullable=False, default="[]")
    estimated_cost = Column(Text, nullable=False, default="{}")
    fallback_reason = Column(Text, nullable=True)
    planner_type = Column(String(30), nullable=False, default="rule_fallback")
    raw_output = Column(Text, nullable=True)
    repair_records = Column(Text, nullable=False, default="[]")
    created_at = Column(DateTime, nullable=False, default=lambda: datetime.now(timezone.utc))
    confirmed_at = Column(DateTime, nullable=True)

    def _get_json(self, field: str, default):
        import json
        value = getattr(self, field, None)
        try:
            return json.loads(value) if value else default
        except (json.JSONDecodeError, TypeError):
            return default

    def set_json(self, field: str, value) -> None:
        import json
        setattr(self, field, json.dumps(value, ensure_ascii=False))

    def to_dict(self, tasks=None) -> dict:
        payload = {
            "id": self.id,
            "plan_id": self.plan_id,
            "goal_id": self.goal_id,
            "version": self.version,
            "status": self.status,
            "task_mode": self.task_mode,
            "goal_summary": self.goal_summary,
            "assumptions": self._get_json("assumptions", []),
            "required_context": self._get_json("required_context", []),
            "activation_reason": self.activation_reason,
            "final_acceptance_criteria": self._get_json("final_acceptance_criteria", []),
            "human_approval_points": self._get_json("human_approval_points", []),
            "estimated_cost": self._get_json("estimated_cost", {}),
            "fallback_reason": self.fallback_reason,
            "planner_type": self.planner_type,
            "raw_output": self.raw_output,
            "repair_records": self._get_json("repair_records", []),
            "created_at": self.created_at.isoformat() if self.created_at else None,
            "confirmed_at": self.confirmed_at.isoformat() if self.confirmed_at else None,
        }
        if tasks is not None:
            payload["tasks"] = [task.to_dict() for task in tasks]
        return payload


class PlanTask(Base):
    """A plan-level task whose client ID remains stable inside one plan version."""

    __tablename__ = "plan_tasks"
    __table_args__ = (
        UniqueConstraint("plan_version_id", "client_task_id", name="uq_plan_task_client_id"),
    )

    id = Column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    plan_version_id = Column(String(36), nullable=False, index=True)
    client_task_id = Column(String(100), nullable=False)
    objective = Column(Text, nullable=False)
    task_type = Column(String(30), nullable=False)
    required_capabilities = Column(Text, nullable=False, default="[]")
    required_tools = Column(Text, nullable=False, default="[]")
    dependencies = Column(Text, nullable=False, default="[]")
    acceptance_criteria = Column(Text, nullable=False, default="[]")
    risk_level = Column(String(20), nullable=False, default="low")
    parallel_safe = Column(Boolean, nullable=False, default=False)
    context_query = Column(Text, nullable=False, default="")
    approval_required = Column(Boolean, nullable=False, default=False)
    workspace_scope = Column(Text, nullable=True)
    merge_strategy = Column(Text, nullable=True)
    runtime_task_id = Column(String(36), nullable=True, index=True)
    source = Column(String(30), nullable=False, default="created")
    created_at = Column(DateTime, nullable=False, default=lambda: datetime.now(timezone.utc))

    def _get_json(self, field: str):
        import json
        try:
            return json.loads(getattr(self, field, None) or "[]")
        except (json.JSONDecodeError, TypeError):
            return []

    def set_json(self, field: str, value) -> None:
        import json
        setattr(self, field, json.dumps(value, ensure_ascii=False))

    def to_dict(self) -> dict:
        return {
            "id": self.id,
            "plan_version_id": self.plan_version_id,
            "client_task_id": self.client_task_id,
            "objective": self.objective,
            "task_type": self.task_type,
            "required_capabilities": self._get_json("required_capabilities"),
            "required_tools": self._get_json("required_tools"),
            "dependencies": self._get_json("dependencies"),
            "acceptance_criteria": self._get_json("acceptance_criteria"),
            "risk_level": self.risk_level,
            "parallel_safe": self.parallel_safe,
            "context_query": self.context_query,
            "approval_required": self.approval_required,
            "workspace_scope": self.workspace_scope,
            "merge_strategy": self.merge_strategy,
            "runtime_task_id": self.runtime_task_id,
            "source": self.source,
            "created_at": self.created_at.isoformat() if self.created_at else None,
        }


class PlanChange(Base):
    """Auditable diff metadata connecting two immutable plan versions."""

    __tablename__ = "plan_changes"

    id = Column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    goal_id = Column(String(36), nullable=False, index=True)
    from_plan_version_id = Column(String(36), nullable=True, index=True)
    to_plan_version_id = Column(String(36), nullable=False, index=True)
    change_type = Column(String(30), nullable=False, default="created")
    reason = Column(Text, nullable=False)
    evidence = Column(Text, nullable=False, default="[]")
    retained_task_ids = Column(Text, nullable=False, default="[]")
    cancelled_task_ids = Column(Text, nullable=False, default="[]")
    added_task_ids = Column(Text, nullable=False, default="[]")
    replaced_task_ids = Column(Text, nullable=False, default="[]")
    created_at = Column(DateTime, nullable=False, default=lambda: datetime.now(timezone.utc))

    def _get_json(self, field: str):
        import json
        try:
            return json.loads(getattr(self, field, None) or "[]")
        except (json.JSONDecodeError, TypeError):
            return []

    def set_json(self, field: str, value) -> None:
        import json
        setattr(self, field, json.dumps(value, ensure_ascii=False))

    def to_dict(self) -> dict:
        return {
            "id": self.id,
            "goal_id": self.goal_id,
            "from_plan_version_id": self.from_plan_version_id,
            "to_plan_version_id": self.to_plan_version_id,
            "change_type": self.change_type,
            "reason": self.reason,
            "evidence": self._get_json("evidence"),
            "retained_task_ids": self._get_json("retained_task_ids"),
            "cancelled_task_ids": self._get_json("cancelled_task_ids"),
            "added_task_ids": self._get_json("added_task_ids"),
            "replaced_task_ids": self._get_json("replaced_task_ids"),
            "created_at": self.created_at.isoformat() if self.created_at else None,
        }


class Artifact(Base):
    __tablename__ = "artifacts"

    id = Column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    run_id = Column(String(36), nullable=True, index=True)
    task_id = Column(String(36), nullable=False, index=True)
    type = Column(String(50), nullable=False, default="file")
    path = Column(Text, nullable=True)
    checksum = Column(String(128), nullable=True)
    verification_status = Column(String(50), nullable=True)
    metadata_json = Column(Text, nullable=False, default="{}")
    created_at = Column(DateTime, nullable=False, default=lambda: datetime.now(timezone.utc))


class VerificationResult(Base):
    __tablename__ = "verification_results"

    id = Column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    task_id = Column(String(36), nullable=False, index=True)
    criterion_type = Column(String(50), nullable=False)
    command_or_rule = Column(Text, nullable=False)
    status = Column(String(50), nullable=False)
    evidence = Column(Text, nullable=True)
    exit_code = Column(Integer, nullable=True)
    created_at = Column(DateTime, nullable=False, default=lambda: datetime.now(timezone.utc))


class WorkspaceCheckpoint(Base):
    __tablename__ = "workspace_checkpoints"

    id = Column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    goal_id = Column(String(36), nullable=False, index=True)
    task_id = Column(String(36), nullable=False, index=True)
    path = Column(Text, nullable=False)
    existed = Column(Boolean, nullable=False, default=False)
    content = Column(Text, nullable=True)
    checksum = Column(String(128), nullable=True)
    created_at = Column(DateTime, nullable=False, default=lambda: datetime.now(timezone.utc))


class WorkspaceWorktree(Base):
    __tablename__ = "workspace_worktrees"

    id = Column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    goal_id = Column(String(36), nullable=False, index=True)
    task_id = Column(String(36), nullable=False, unique=True, index=True)
    path = Column(Text, nullable=False, unique=True)
    base_ref = Column(String(255), nullable=False, default="HEAD")
    base_commit_sha = Column(String(64), nullable=True)
    branch_name = Column(String(255), nullable=True, unique=True)
    agent_id = Column(String(36), nullable=True, index=True)
    commit_sha = Column(String(64), nullable=True)
    merge_commit_sha = Column(String(64), nullable=True)
    merge_output = Column(Text, nullable=True)
    conflict_files = Column(Text, nullable=False, default="[]")
    status = Column(String(30), nullable=False, default="active")
    created_at = Column(DateTime, nullable=False, default=lambda: datetime.now(timezone.utc))
    merged_at = Column(DateTime, nullable=True)
    removed_at = Column(DateTime, nullable=True)

    def to_dict(self) -> dict:
        try:
            conflicts = json.loads(self.conflict_files or "[]")
        except json.JSONDecodeError:
            conflicts = []
        return {
            "id": self.id, "goal_id": self.goal_id, "task_id": self.task_id,
            "agent_id": self.agent_id, "path": self.path, "base_ref": self.base_ref,
            "base_commit_sha": self.base_commit_sha, "branch_name": self.branch_name,
            "commit_sha": self.commit_sha, "merge_commit_sha": self.merge_commit_sha,
            "merge_output": self.merge_output, "conflict_files": conflicts,
            "status": self.status,
            "created_at": self.created_at.isoformat() if self.created_at else None,
            "merged_at": self.merged_at.isoformat() if self.merged_at else None,
            "removed_at": self.removed_at.isoformat() if self.removed_at else None,
        }


class RuntimeRun(Base):
    __tablename__ = "runtime_runs"

    id = Column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    goal_id = Column(String(36), nullable=False, index=True)
    execution_mode = Column(String(20), nullable=False)
    status = Column(String(30), nullable=False, default="running")
    started_at = Column(DateTime, nullable=False, default=lambda: datetime.now(timezone.utc))
    ended_at = Column(DateTime, nullable=True)
    final_verification_status = Column(String(50), nullable=True)
    budget_tokens = Column(Integer, nullable=False, default=100000)
    budget_cost_usd = Column(Float, nullable=True)
    max_duration_seconds = Column(Integer, nullable=False, default=3600)


@event.listens_for(Goal, "before_update")
@event.listens_for(Task, "before_update")
def receive_before_update(mapper, connection, target):
    target.updated_at = datetime.now(timezone.utc)
